sklego/meta.py

```diff
@@ -114,62 +228,129 @@ def fit(self, X, y):
+            pd.Series(self.__predict_group(X, level_columns)) for level_columns in self.group_colnames_hierarchical_
+        ], axis=1)
+
+        # This is a Series with values the arrays
+        prediction_groups = X[self.group_colnames_].agg(func=tuple, axis=1)
+
+        shrinkage_factors = prediction_groups.map(self.shrinkage_factors_)
+
+        if any(shrinkage_factors.isnull()):
+            # Handle new groups
+            diff = set(prediction_groups) - set(self.shrinkage_factors_.keys())
+
+            raise ValueError(f"found a group(s) {diff} in `.predict` that was not in `.fit`")
+
+        # So convert it to a dataframe
+        shrinkage_factors = pd.DataFrame.from_dict(shrinkage_factors.to_dict()).T
+
+        return (hierarchical_predictions * shrinkage_factors).sum(axis=1)
+
     def predict(self, X):
         """
-        Predict new data by making random guesses.
+        Predict on new data.
 
         :param X: array-like, shape=(n_columns, n_samples,) training data.
         :return: array, shape=(n_samples,) the predicted data
         """
         self.__validate(X)
 
         check_is_fitted(self, ['estimators_', 'groups_', 'group_colnames_',
-                               'X_colnames_', 'fallback_'])
+                               'value_colnames_', 'fallback_'])
         if isinstance(X, np.ndarray):
             X = pd.DataFrame(X, columns=[str(_) for _ in range(X.shape[1])])
 
         if any([c not in X.columns for c in self.group_colnames_]):
             raise ValueError(f"group columns {self.group_colnames_} not in {X.columns}")
-        if any([c not in X.columns for c in self.X_colnames_]):
-            raise ValueError(f"columns to use {self.X_colnames_} not in {X.columns}")
+        if any([c not in X.columns for c in self.value_colnames_]):
+            raise ValueError(f"columns to use {self.value_colnames_} not in {X.columns}")
 
-        try:
-            return (X
-                    .groupby(self.group_colnames_, as_index=False)
-                    .apply(lambda d: pd.DataFrame(
-                        self.estimators_.get(d.name, self.fallback_).predict(d[self.X_colnames_]), index=d.index))
-                    .values
-                    .squeeze())
-        except AttributeError:
-            culprits = set(pd.concat([X[self.group_colnames_].drop_duplicates().assign(new=1),
-                                      self.groups_.assign(new=0)])
-                           .drop_duplicates()
-                           .loc[lambda d: d['new'] == 1]
-                           .itertuples())
-            raise ValueError(f"found a group(s) {culprits} in `.predict` that was not in `.fit`")
+        if not self.shrinkage:
+            return self.__predict_group(X, group_colnames=self.group_colnames_)
+        else:
+            return self.__predict_shrinkage_groups(X)
+
+    def __expanding_list(self, list_to_extent: List[str], return_type=list) -> Union[List[list], List[tuple]]:
```

Comment:
Separate function or `@staticmethod`?

Reply:
It might be preferable to move such a function to `sklego.common`.

Reply:
I'll move it to `sklego.common`